File: js/nova/datastruct/list/LinkedList.js
var LinkedList = function () {
	this.start = nova_null;
	this.current = nova_null;
	this.size = 0;
	
	
};

LinkedList.prototype.LinkedList = function () {
	this.this();
};

LinkedList.prototype.destroy = function () {
};

LinkedList.prototype.addAll = function (data) {
	var nova_local_0;
	var d;
	nova_local_0 = (data).iterator();
	while (nova_local_0.hasNext()) {
		d = nova_local_0.next();
		this.add(d);
	}
	return this;
};

LinkedList.prototype.add = function (data) {
	var node;
	node = new ListNode(data);
	if (this.start == null) {
		this.start = node;
		this.current = node;
	} else {
		this.current.next = node;
	}
	this.current = node;
	this.size++;
	return this;
};

LinkedList.prototype.remove = function (data) {
	var prev;
	var cur;
	if (this.start.data == data) {
		this.start = this.start.next;
	}
	prev = this.start;
	cur = this.start.next;
	while (cur != null) {
		var d;
		d = cur.data;
		if (d == data) {
			prev.next = cur.next;
			this.size--;
		}
		cur = cur.next;
	}
	return this;
};

LinkedList.prototype.contains = function (value) {
	var contextArg17;
	return this.any(this.testLambda17);
};

LinkedList.prototype.toArray = function () {
	var array;
	var i;
	var nova_local_0;
	var element;
	array = new Array(this.size);
	i = 0;
	nova_local_0 = (this).iterator();
	while (nova_local_0.hasNext()) {
		element = nova_local_0.next();
		array.set(i++, element);
	}
	return array;
};

LinkedList.prototype.map = function (mapFunc) {
	var array;
	var i;
	var nova_local_0;
	var element;
	array = new LinkedList();
	i = 0;
	nova_local_0 = (this).iterator();
	while (nova_local_0.hasNext()) {
		element = nova_local_0.next();
		array.add(mapFunc(element, i++, this));
	}
	return array;
};

LinkedList.prototype.forEach = function (func) {
	var i;
	var nova_local_0;
	var element;
	i = 0;
	nova_local_0 = (this).iterator();
	while (nova_local_0.hasNext()) {
		element = nova_local_0.next();
		func(element, i++, this);
	}
};

LinkedList.prototype.any = function (anyFunc) {
	var nova_local_0;
	var element;
	nova_local_0 = (this).iterator();
	while (nova_local_0.hasNext()) {
		element = nova_local_0.next();
		if (anyFunc(element)) {
			return true;
		}
	}
	return false;
};

LinkedList.prototype.all = function (allFunc) {
	var nova_local_0;
	var element;
	nova_local_0 = (this).iterator();
	while (nova_local_0.hasNext()) {
		element = nova_local_0.next();
		if (!allFunc(element)) {
			return false;
		}
	}
	return true;
};

LinkedList.prototype.filter = function (filterFunc) {
	var filtered;
	var i;
	var nova_local_0;
	var element;
	filtered = new LinkedList();
	i = 0;
	nova_local_0 = (this).iterator();
	while (nova_local_0.hasNext()) {
		element = nova_local_0.next();
		if (filterFunc(element, i++, this)) {
			filtered.add(element);
		}
	}
	return filtered;
};

LinkedList.prototype.take = function (howMany) {
	var list;
	var nova_local_0;
	var element;
	if (howMany > this.size) {
		howMany = this.size;
	}
	list = new LinkedList();
	nova_local_0 = (this).iterator();
	while (nova_local_0.hasNext()) {
		element = nova_local_0.next();
		if (list.size == howMany) {
			break;
		}
		list.add(element);
	}
	return list;
};

LinkedList.prototype.skip = function (howMany) {
	var list;
	var i;
	var nova_local_0;
	var element;
	list = new Array();
	i = 0;
	nova_local_0 = (this).iterator();
	while (nova_local_0.hasNext()) {
		element = nova_local_0.next();
		if (i++ > howMany) {
			list.add(element);
		}
	}
	return list;
};

LinkedList.prototype.firstWhere = function (func) {
	var nova_local_0;
	var element;
	nova_local_0 = (this).iterator();
	while (nova_local_0.hasNext()) {
		element = nova_local_0.next();
		if (func(element)) {
			return element;
		}
	}
	return null;
};

LinkedList.prototype.reverse = function () {
	var list;
	var prev;
	var current;
	var next;
	list = new LinkedList();
	prev = null;
	current = null;
	next = null;
	if (this.start != null) {
		current = this.start.clone();
	}
	while (current != null) {
		next = current.next;
		current.next = null;
		if (next != null) {
			next = next.clone();
		}
		if (prev != null) {
			current.next = prev.clone();
		}
		prev = current;
		current = next;
	}
	list.start = prev;
	return list;
};

LinkedList.prototype.join = function (delimiter) {
	var str;
	var passed;
	var nova_local_0;
	var element;
	str = "";
	passed = false;
	nova_local_0 = (this).iterator();
	while (nova_local_0.hasNext()) {
		element = nova_local_0.next();
		if (passed) {
			str = str.concat(delimiter);
		} else {
			passed = true;
		}
		str = str.concat(element.toString());
	}
	return str;
};

LinkedList.prototype.this = function () {
};

LinkedList.prototype.testLambda17 = function (_1) {
	return _1 == value;
};

LinkedList.prototype.accessor_iterator = function () {
	return new LinkedListIterator(this);
};

LinkedList.prototype.mutator_iterator = function () {
};

LinkedList.prototype.accessor_first = function () {
	return this.start;
};

LinkedList.prototype.mutator_first = function () {
};

LinkedList.prototype.accessor_last = function () {
	return this.current;
};

LinkedList.prototype.mutator_last = function () {
};

LinkedList.prototype.super = function () {
};

File: java/nova/datastruct/list/LinkedList.java
package nova.datastruct.list;

import java.util.Optional;
import nova.exception.ExceptionData;
import nova.exception.Exception;
import nova.exception.DivideByZeroException;
import nova.io.NovaConsole;
import nova.primitive.number.Number;
import nova.primitive.number.NovaByte;
import nova.primitive.number.NovaShort;
import nova.primitive.number.NovaInt;
import nova.primitive.number.NovaLong;
import nova.primitive.number.NovaFloat;
import nova.primitive.number.NovaDouble;
import nova.primitive.Null;
import nova.primitive.number.Char;
import nova.primitive.Bool;
import nova.datastruct.list.NovaArray;
import nova.datastruct.list.IntArray;
import nova.datastruct.list.CharArray;
import nova.datastruct.list.DoubleArray;
import nova.datastruct.list.IntRange;
import nova.thread.Thread;
import nova.thread.async.Async;
import nova.gc.GC;
import nova.math.NovaMath;
import nova.NovaObject;
import nova.NovaString;
import nova.System;
import nova.Class;
import nova.datastruct.list.ArrayIterator;
import nova.datastruct.list.LinkedListIterator;
import nova.datastruct.list.List;
import nova.datastruct.list.ListNode;

public class LinkedList extends NovaObject implements List
{
	private ListNode start;
	private ListNode current;
	
	public int size;
	
	
	
	
	public LinkedList()
	{
		init();
	}
	
	public LinkedListIterator accessor_iterator()
	{
		return new LinkedListIterator(this);
	}
	
	private LinkedListIterator mutator_iterator()
	{
	}
	
	public ListNode accessor_first()
	{
		return start;
	}
	
	private ListNode mutator_first()
	{
	}
	
	public ListNode accessor_last()
	{
		return current;
	}
	
	private ListNode mutator_last()
	{
	}
	
	public LinkedList addAll(NovaArray data)
	{
		ArrayIterator nova_local_0;
		NovaObject d;
		nova_local_0 = (data).accessor_iterator();
		while (nova_local_0.accessor_hasNext())
		{
			d = nova_local_0.accessor_next();
			add(d);
		}
		return this;
	}
	
	public LinkedList add(NovaObject data)
	{
		ListNode node;
		node = new ListNode(data);
		if (start == null)
		{
			start = node;
			current = node;
		}
		else
		{
			current.next = node;
		}
		current = node;
		size++;
		return this;
	}
	
	public LinkedList remove(NovaObject data)
	{
		ListNode prev;
		ListNode cur;
		if (start.data == data)
		{
			start = start.next;
		}
		prev = start;
		cur = start.next;
		while (cur != null)
		{
			NovaObject d;
			d = cur.data;
			if (d == data)
			{
				prev.next = cur.next;
				size--;
			}
			cur = cur.next;
		}
		return this;
	}
	
	public boolean contains(NovaObject value)
	{
		return any(testLambda32);
	}
	
	public NovaArray toArray()
	{
		NovaArray array;
		int i;
		LinkedListIterator nova_local_0;
		NovaObject element;
		array = new NovaArray(size);
		i = 0;
		nova_local_0 = (this).accessor_iterator();
		while (nova_local_0.accessor_hasNext())
		{
			element = nova_local_0.accessor_next();
			array.set(i++, element);
		}
		return array;
	}
	
	public LinkedList map(NovaUtilities.Function3<E, Int, LinkedList, Out> mapFunc)
	{
		LinkedList array;
		int i;
		LinkedListIterator nova_local_0;
		NovaObject element;
		array = new LinkedList();
		i = 0;
		nova_local_0 = (this).accessor_iterator();
		while (nova_local_0.accessor_hasNext())
		{
			element = nova_local_0.accessor_next();
			array.add(mapFunc.call(element, i++, this));
		}
		return array;
	}
	
	public LinkedList forEach(NovaUtilities.Consumer3<E, Int, LinkedList> func)
	{
		int i;
		LinkedListIterator nova_local_0;
		NovaObject element;
		i = 0;
		nova_local_0 = (this).accessor_iterator();
		while (nova_local_0.accessor_hasNext())
		{
			element = nova_local_0.accessor_next();
			func.call(element, i++, this);
		}
		return this;
	}
	
	public boolean any(NovaUtilities.Function1<E, Bool> anyFunc)
	{
		LinkedListIterator nova_local_0;
		NovaObject element;
		nova_local_0 = (this).accessor_iterator();
		while (nova_local_0.accessor_hasNext())
		{
			element = nova_local_0.accessor_next();
			if (anyFunc.call(element))
			{
				return true;
			}
		}
		return false;
	}
	
	public boolean all(NovaUtilities.Function1<E, Bool> allFunc)
	{
		LinkedListIterator nova_local_0;
		NovaObject element;
		nova_local_0 = (this).accessor_iterator();
		while (nova_local_0.accessor_hasNext())
		{
			element = nova_local_0.accessor_next();
			if (!allFunc.call(element))
			{
				return false;
			}
		}
		return true;
	}
	
	public LinkedList filter(NovaUtilities.Function3<E, Int, LinkedList, Bool> filterFunc)
	{
		LinkedList filtered;
		int i;
		LinkedListIterator nova_local_0;
		NovaObject element;
		filtered = new LinkedList();
		i = 0;
		nova_local_0 = (this).accessor_iterator();
		while (nova_local_0.accessor_hasNext())
		{
			element = nova_local_0.accessor_next();
			if (filterFunc.call(element, i++, this))
			{
				filtered.add(element);
			}
		}
		return filtered;
	}
	
	public LinkedList take(int howMany)
	{
		LinkedList list;
		LinkedListIterator nova_local_0;
		NovaObject element;
		if (howMany > size)
		{
			howMany = size;
		}
		list = new LinkedList();
		nova_local_0 = (this).accessor_iterator();
		while (nova_local_0.accessor_hasNext())
		{
			element = nova_local_0.accessor_next();
			if (list.size == howMany)
			{
				break;
			}
			list.add(element);
		}
		return list;
	}
	
	public LinkedList skip(int howMany)
	{
		NovaArray list;
		int i;
		LinkedListIterator nova_local_0;
		NovaObject element;
		list = new NovaArray();
		i = 0;
		nova_local_0 = (this).accessor_iterator();
		while (nova_local_0.accessor_hasNext())
		{
			element = nova_local_0.accessor_next();
			if (i++ > howMany)
			{
				list.add(element);
			}
		}
		return list;
	}
	
	public NovaObject firstWhere(NovaUtilities.Function1<E, Bool> func)
	{
		LinkedListIterator nova_local_0;
		NovaObject element;
		nova_local_0 = (this).accessor_iterator();
		while (nova_local_0.accessor_hasNext())
		{
			element = nova_local_0.accessor_next();
			if (func.call(element))
			{
				return element;
			}
		}
		return null;
	}
	
	public LinkedList reverse()
	{
		LinkedList list;
		ListNode prev;
		ListNode current;
		ListNode next;
		list = new LinkedList();
		prev = null;
		current = null;
		next = null;
		if (start != null)
		{
			current = start.clone();
		}
		while (current != null)
		{
			next = current.next;
			current.next = null;
			if (next != null)
			{
				next = next.clone();
			}
			if (prev != null)
			{
				current.next = prev.clone();
			}
			prev = current;
			current = next;
		}
		list.start = prev;
		return list;
	}
	
	public void init()
	{
	}
	
	private static boolean testLambda32(NovaObject _1)
	{
		return _1 == value;
	}
	
}
